File: Scrabbler.Tests/BoardCorrectionParserTests.cs
using Scrabbler.Domain.BoardModel;

namespace Scrabbler.Tests;

public sealed class BoardCorrectionParserTests
{
    [Fact]
    public void AppliesSingleLetterCorrection()
    {
        var board = EmptyBoard();

        board = BoardCorrectionParser.ApplyCorrections(board, "A1=Ł");

        Assert.Equal('Ł', board[0, 0].Letter);
        Assert.False(board[0, 0].IsBlank);
    }

    [Fact]
    public void AppliesKnownBlankTileCorrection()
    {
        var board = EmptyBoard();

        board = BoardCorrectionParser.ApplyCorrections(board, "H8=Ł?");

        Assert.Equal('Ł', board[7, 7].Letter);
        Assert.True(board[7, 7].IsBlank);
    }

    [Fact]
    public void ClearsCellWithDotOrQuestionMark()
    {
        var board = EmptyBoard().SetCell(7, 7, 'A');

        Assert.Null(BoardCorrectionParser.ApplyCorrections(board, "H8=.")[7, 7].Letter);
        Assert.Null(BoardCorrectionParser.ApplyCorrections(board, "H8=?")[7, 7].Letter);
    }

    [Fact]
    public void AppliesCommaSeparatedCorrections()
    {
        var board = EmptyBoard();

        board = BoardCorrectionParser.ApplyCorrections(board, "A1=A, B2=Ń, C3=Ż?");

        Assert.Equal('A', board[0, 0].Letter);
        Assert.Equal('Ń', board[1, 1].Letter);
        Assert.Equal('Ż', board[2, 2].Letter);
        Assert.True(board[2, 2].IsBlank);
    }

    [Theory]
    [InlineData("P1=A")]
    [InlineData("A16=A")]
    [InlineData("A1=1")]
    [InlineData("A1")]
    public void RejectsInvalidCorrections(string correction)
    {
        Assert.Throws<ArgumentException>(() => BoardCorrectionParser.ApplyCorrections(EmptyBoard(), correction));
    }

    private static Board EmptyBoard()
    {
        return new Board(new BonusType[Board.Size, Board.Size]);
    }
}

File: Scrabbler.iOS/ScrabblerKit/Tests/ScrabblerKitTests/BoardCorrectionParserTests.swift
import Testing
@testable import ScrabblerKit

@Suite("Board correction parser")
struct BoardCorrectionParserTests {
    @Test func appliesLetterCorrection() throws {
        let board = try BoardCorrectionParser.applyCorrections(to: emptyBoard(), input: "A1=Ł")

        #expect(board[0, 0].letter == "Ł")
        #expect(!board[0, 0].isBlank)
    }

    @Test func appliesBlankCorrection() throws {
        let board = try BoardCorrectionParser.applyCorrections(to: emptyBoard(), input: "H8=Ż?")

        #expect(board[7, 7].letter == "Ż")
        #expect(board[7, 7].isBlank)
    }

    @Test func clearsCellWithDotOrQuestionMark() throws {
        let board = emptyBoard().setCell(row: 9, column: 9, letter: "A")

        #expect(try BoardCorrectionParser.applyCorrections(to: board, input: "J10=.")[9, 9].letter == nil)
        #expect(try BoardCorrectionParser.applyCorrections(to: board, input: "J10=?")[9, 9].letter == nil)
    }

    @Test func appliesCommaSeparatedCorrections() throws {
        let board = try BoardCorrectionParser.applyCorrections(to: emptyBoard(), input: "A1=Ł, H8=?, J10=Ń")

        #expect(board[0, 0].letter == "Ł")
        #expect(board[7, 7].letter == nil)
        #expect(board[9, 9].letter == "Ń")
    }

    @Test func formatsCorrectionEntriesAndAppendsThem() {
        #expect(BoardCorrectionParser.correctionEntry(row: 0, column: 0) == "A1=")
        #expect(BoardCorrectionParser.correctionEntry(row: 7, column: 7) == "H8=")

        #expect(
            BoardCorrectionParser.appendCorrectionEntries(
                to: "",
                coordinates: [(row: 0, column: 0), (row: 7, column: 7)]
            ) == "A1=, H8="
        )
        #expect(
            BoardCorrectionParser.appendCorrectionEntries(
                to: "B2=A",
                coordinates: [(row: 9, column: 9)]
            ) == "B2=A, J10="
        )
        #expect(
            BoardCorrectionParser.appendCorrectionEntries(
                to: "B2=A",
                coordinates: []
            ) == "B2=A"
        )
    }

    @Test func replacesLastMatchingCorrectionValue() {
        #expect(
            BoardCorrectionParser.replaceLastCorrectionValue(
                in: "A1=Ł, H8=, A1=",
                coordinate: "A1",
                value: "Ń"
            ) == "A1=Ł, H8=, A1=Ń"
        )
        #expect(
            BoardCorrectionParser.replaceLastCorrectionValue(
                in: "A1=Ł",
                coordinate: "H8",
                value: "."
            ) == "A1=Ł, H8=."
        )
        #expect(
            BoardCorrectionParser.replaceLastCorrectionValue(
                in: "",
                coordinate: "h8",
                value: "Ż?"
            ) == "H8=Ż?"
        )
    }

    @Test func extractsCorrectionCellKeys() throws {
        let keys = try BoardCorrectionParser.correctionCellKeys(from: "A1=Ł, H8=, J10=.")

        #expect(keys == [
            BoardCorrectionParser.cellKey(row: 0, column: 0),
            BoardCorrectionParser.cellKey(row: 7, column: 7),
            BoardCorrectionParser.cellKey(row: 9, column: 9)
        ])
    }

    @Test func rejectsInvalidCoordinate() throws {
        #expect(throws: ScrabblerError.self) {
            _ = try BoardCorrectionParser.applyCorrections(to: emptyBoard(), input: "Q1=A")
        }
    }

    @Test func rejectsInvalidLetter() throws {
        #expect(throws: ScrabblerError.self) {
            _ = try BoardCorrectionParser.applyCorrections(to: emptyBoard(), input: "A1=X")
        }
    }

    private func emptyBoard() -> Board {
        Board(bonuses: Array(repeating: Array(repeating: BonusType.none, count: Board.size), count: Board.size))
    }
}
